### DISCUSSION TITLE
Make the rtsp server only "read" the stream being broadcast to it on demand (i.e if someone is trying to fetch the stream from the server)

### DISCUSSION THREAD

**Discussion by Mohamedemad4** [bluenviron/mediamtx#275] [Questions]
## Which version are you using?
v0.13.2
OS: Linux (Ubuntu 20.04)

## Question
- Make the rtsp Server only "read" the stream being broadcasted to it if someone is requesting to read that stream from the server
- does ```sourceOnDemand``` work if rtsp-simple-server is deployed in server mode? and if not what's an alternative? 

## Description
I have a raspberry pi streaming a webcam to an RTSP server running rtsp-simple-server with ffmpeg
- the RPI is using this ffmpeg cmd
   - ```ffmpeg -f v4l2 -i /dev/video0  -maxrate 4M -b 2M -f rtsp rtsp://user:pass@192.168.1.3:8554/stream```
- What i would like to do is to have the server **tell ffmpeg** to "stop streaming" if no one is reading the rtsp stream from the server (to **save on bandwidth between the rpi and the server**)
- since this setup should be running 24/7 and an rtsp client should only be requesting the rtsp stream throughout the day and not 24/7.
- i have tried setting ```sourceOnDemand``` to ```yes``` in all paths. but the server is sill receiving the same bandwidth weather somone is reading the rtsp stream or not 

Also if i am going about this the wrong way. please do tell me how this can be achieved/where to look. Or if i can make a PR or something and **contribute**  this functionality (idk a lot about rtsp if that wasn't obvious lol) 

@aler9

**Comment by aler9** [comment]
Hello, this can be solved in a lot ot ways, these are the two that come in mind

## way 1

Start a second instance of rtsp-simple-server on the Raspberry Pi with this configuration

```yml
paths:
  cam:
    runOnInit: ffmpeg -f v4l2 -i /dev/video0 -maxrate 4M -b 2M -f rtsp rtsp://localhost:8554/cam
```

Edit the configuration of the main instance running on the server in this way

```yml
paths:
  cam:
    source: rtsp://raspberry-url/cam
    sourceOnDemand: yes
```

The setup is now this
camera --> first rtsp server (running locally) --> 2nd rtsp server (running on the remote server)

In this way, the stream will be pulled only when requested, saving bandwidth.

## way 2

edit the configuration on the server in this way

```yml
paths:
  stream:
    runOnDemand: /script.sh
```

create a script with this content:

```sh
#!/bin/sh

trap 'echo sending ssh command to the raspberry pi to stop the stream' INT

echo sending ssh command to the raspberry pi to start the stream

sleep 86400
```

and edit the two parts to do exactly what they are supposed to do :)

**Comment by Mohamedemad4** [comment]
in my setup the rapsberry pi is behind a NAT so ways 1 and 2 aren't really an option.
I thought about writing a simple bash script that polls a REST server to ask for permission to stream. But that just seems inefficient and a bad solution overall.
Surely there has to be a way to implement this over the RTSP protocol. Right? even if it means i will have to open a PR.

**Comment by dlueth** [comment]
@Mohamedemad4 no idea if it's works in this Szenario but you could try RabbitMQ. OnDemand command would be a shell script on the rtsp-simple server host that produces one message on start and another (different) one on trap INT. You would need to consume these on your RPI and act accordingly.

**Comment by Mohamedemad4** [comment]
thanks for the suggestion @dlueth. what we ended up doing was quiet similar. Just using redis Pub/Sub instead. I thought since RTSP can be used for controlling remote media (from my understanding at least) then something like that should have been trivial to implement **inside** RTSP. I am closing this issue but, if anyone has any ideas to do this inside RTSP feel free to re-open.

**Comment by aler9** [comment]
RTSP can be used only for unidirectional communication from client to server, not the opposite - a MQTT-based solution is the best approach in this case.

[a comment by an automated account is not shown]
